drivers/sensor/mhz19b/mhz19b.c

```diff
@@ -0,0 +1,340 @@
+{
+	struct mhz19b_data *data = dev->data;
+	uint8_t checksum;
+	int ret;
+
+	ret = mhz19b_send_cmd(dev, cmd_idx, true);
+	if (ret < 0) {
+		return ret;
+	}
+
+	checksum = mhz19b_checksum(data->rd_data);
+	if (checksum != data->rd_data[MHZ19B_CHECKSUM_IDX]) {
+		data->data_valid = false;
+		return -EBADMSG;
+	}
+
+	data->data_valid = true;
+
+	switch (cmd_idx) {
+	case MHZ19B_CMD_IDX_GET_CO2:
+		data->data = sys_get_be16(&data->rd_data[2]);
+		break;
+	case MHZ19B_CMD_IDX_GET_RANGE:
+		data->data = sys_get_be16(&data->rd_data[4]);
+		break;
+	case MHZ19B_CMD_IDX_GET_ABC:
+		data->data = (uint16_t)data->rd_data[7];
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int mhz19b_channel_get(const struct device *dev, enum sensor_channel chan,
+			      struct sensor_value *val)
+{
+	struct mhz19b_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	if (!data->data_valid) {
+		LOG_DBG("Checksum mismatch");
+		return -EINVAL;
+	}
+
+	val->val1 = (int32_t)data->data;
+	val->val2 = 0;
+
+	return 0;
+}
+
+static int mhz19b_attr_full_scale_cfg(const struct device *dev, const int range)
+{
+	switch (range) {
+	case 2000:
+		LOG_DBG("Configure range to %d", range);
+		return mhz19b_send_config(dev, MHZ19B_CMD_IDX_SET_RANGE_2000);
+	case 5000:
+		LOG_DBG("Configure range to %d", range);
+		return mhz19b_send_config(dev, MHZ19B_CMD_IDX_SET_RANGE_5000);
+	case 10000:
+		LOG_DBG("Configure range to %d", range);
+		return mhz19b_send_config(dev, MHZ19B_CMD_IDX_SET_RANGE_10000);
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int mhz19b_attr_abc_cfg(const struct device *dev, const bool on)
+{
+	if (on) {
+		LOG_DBG("%s ABC", "Enable");
+		return mhz19b_send_config(dev, MHZ19B_CMD_IDX_SET_ABC_ON);
+	}
+
+	LOG_DBG("%s ABC", "Disable");
+	return mhz19b_send_config(dev, MHZ19B_CMD_IDX_SET_ABC_OFF);
+}
+
+static int mhz19b_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, const struct sensor_value *val)
+{
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return mhz19b_attr_full_scale_cfg(dev, val->val1);
+
+	case SENSOR_ATTR_MHZ19B_ABC:
+		return mhz19b_attr_abc_cfg(dev, val->val1);
+
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int mhz19b_attr_get(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr, struct sensor_value *val)
+{
+	struct mhz19b_data *data = dev->data;
+	int ret;
+
+	if (chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		ret = mhz19b_poll_data(dev, MHZ19B_CMD_IDX_GET_RANGE);
+		break;
+	case SENSOR_ATTR_MHZ19B_ABC:
+		ret = mhz19b_poll_data(dev, MHZ19B_CMD_IDX_GET_ABC);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	val->val1 = (int32_t)data->data;
+	val->val2 = 0;
+
+	return ret;
+}
+
+static int mhz19b_trigger_set(const struct device *dev, const struct sensor_trigger *trig,
```

Comment:
this function is not reguired, sensor API already returns `-ENOSYS` if not implemented.

Reply:
Didn't know this, thanks!